primitive-types/src/lib.rs

```diff
@@ -136,6 +149,39 @@ impl From<U512> for U256 {
 	}
 }
 
+impl From<U512> for U128 {
+	fn from(value: U512) -> U128 {
+		let U512(ref arr) = value;
+		if arr[2] | arr[3] | arr[4] | arr[5] | arr[6] | arr[7] != 0 {
+			panic!("From<U512> for U128: encountered overflow")
+		}
+		let mut ret = [0; 2];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		U128(ret)
+	}
+}
+
+impl From<U128> for U512 {
+	fn from(value: U128) -> U512 {
+		let U128(ref arr) = value;
+		let mut ret = [0; 8];
+		ret[0] = arr[0];
+		ret[1] = arr[1];
+		U512(ret)
+	}
+}
+
+impl From<U128> for U256 {
```

Comment:
For other types it seems that we also have conversions from references (see below), should we just implement for every `T: AsRef<U128>` or sth?

Reply:
I haven't found a way to make it work -- I get conflicting impl if I try to put T inside `From`. I think it's only possible to implement one `From<T>` for a specific type, regardless of the where clause.